File: apps/backend/src/main/java/com/habittracker/model/HabitEntity.java
package com.habittracker.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigInteger;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(
    name = "habits",
    indexes = {
      @Index(name = "habits_user_updated_cursor_idx", columnList = "userId,updatedAt,id")
    }
)
public class HabitEntity extends PanacheEntityBase {
  @Id
  @Column(nullable = false)
  public String id;

  @Column(name = "userId", nullable = false)
  public String userId;

  @Column(nullable = false)
  public String name;

  @Column
  public String description;

  @Column(nullable = false)
  public String color;

  @Column(nullable = false)
  public String icon;

  @Column(nullable = false)
  public String frequency;

  @Column(name = "customDays", columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  public String customDays;

  @Column(name = "schedule", columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  public String schedule;

  @Column(name = "targetStreak", nullable = false)
  public int targetStreak;

  @Column(name = "dailyTarget", nullable = false)
  public int dailyTarget;

  @Column(columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  public String tags;

  @Column(nullable = false)
  public boolean archived;

  @Column(name = "sortOrder", nullable = false)
  public BigInteger sortOrder;

  @Column(name = "reminderTime")
  public String reminderTime;

  @Column(name = "reminderEnabled", nullable = false)
  public boolean reminderEnabled;

  @Column(name = "lastReminderSentAt")
  public Instant lastReminderSentAt;

  @Column(nullable = false)
  public String type;

  @Column(name = "freezeDays", columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  public String freezeDays;

  @Column(name = "createdAt", nullable = false)
  public Instant createdAt;

  @Column(name = "updatedAt", nullable = false)
  public Instant updatedAt;

  @Column(nullable = false)
  public int version;

  @PrePersist
  void prePersist() {
    if (id == null || id.isBlank()) {
      id = UUID.randomUUID().toString();
    }
    if (dailyTarget < 1) {
      dailyTarget = 1;
    }
    if (sortOrder == null) {
      sortOrder = BigInteger.ZERO;
    }
    if (type == null || type.isBlank()) {
      type = "positive";
    }
    if (freezeDays == null) {
      freezeDays = "[]";
    }
    if (version < 1) {
      version = 1;
    }
    if (createdAt == null) {
      createdAt = Instant.now();
    }
    if (updatedAt == null) {
      updatedAt = createdAt;
    }
  }
}

File: apps/web/src/lib/storage/habitEntity.ts
import type { Habit } from '@/types/habit';
import type { HabitEntity } from './db';
import { normalizeSchedule, scheduleFromLegacy } from '@habbit-runner/shared';

export function normalizeStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value.filter((entry): entry is string => typeof entry === 'string');
}

export function normalizeNumberArray(value: unknown): number[] | undefined {
  if (!Array.isArray(value)) {
    return undefined;
  }

  return value.filter((entry): entry is number => typeof entry === 'number' && Number.isFinite(entry));
}

export function normalizeCompletions(value: unknown): Record<string, number> {
  if (!value || typeof value !== 'object') {
    return {};
  }

  const entries = Object.entries(value).flatMap(([date, count]) => {
    if (typeof count !== 'number' || !Number.isFinite(count)) {
      return [];
    }

    return [[date, Math.max(0, Math.trunc(count))] as const];
  });

  return Object.fromEntries(entries);
}

export function habitEntityToDomain(entity: HabitEntity): Habit {
  const customDays = normalizeNumberArray(entity.customDays);

  return {
    id: entity.id,
    name: entity.name,
    description: entity.description ?? '',
    color: entity.color as Habit['color'],
    icon: entity.icon,
    frequency: entity.frequency as Habit['frequency'],
    customDays,
    targetStreak: entity.targetStreak,
    dailyTarget: Math.max(1, Math.trunc(entity.dailyTarget ?? 1)),
    tags: normalizeStringArray(entity.tags),
    completions: normalizeCompletions(entity.completions),
    freezeDays: normalizeStringArray(entity.freezeDays),
    sortOrder: entity.sortOrder ?? Date.parse(entity.createdAt),
    reminderTime: entity.reminderTime ?? undefined,
    reminderEnabled: entity.reminderEnabled ?? true,
    createdAt: entity.createdAt,
    updatedAt: entity.updatedAt,
    version: entity.version,
    archived: entity.archived ?? false,
    type: entity.type ?? 'positive',
    schedule:
      normalizeSchedule(entity.schedule) ??
      scheduleFromLegacy(entity.frequency as Habit['frequency'], customDays)
  };
}